river/InputRelay.zig

```diff
@@ -0,0 +1,225 @@
+
+    const text_input = self.getFocusedTextInput() orelse return;
+    if (text_input.wlr_text_input.focused_surface) |surface| {
+        text_input.setPendingFocusedSurface(surface);
+    }
+    text_input.wlr_text_input.sendLeave();
+}
+
+fn handleInputMethodGrabKeyboard(
+    listener: *wl.Listener(*wlr.InputMethodV2.KeyboardGrab),
+    keyboard_grab: *wlr.InputMethodV2.KeyboardGrab,
+) void {
+    const self = @fieldParentPtr(Self, "grab_keyboard", listener);
+
+    const active_keyboard = self.seat.wlr_seat.getKeyboard();
+    keyboard_grab.setKeyboard(active_keyboard);
+
+    keyboard_grab.events.destroy.add(&self.grab_keyboard_destroy);
+}
+
+fn handleInputMethodGrabKeyboardDestroy(
+    listener: *wl.Listener(*wlr.InputMethodV2.KeyboardGrab),
+    keyboard_grab: *wlr.InputMethodV2.KeyboardGrab,
+) void {
+    const self = @fieldParentPtr(Self, "grab_keyboard_destroy", listener);
+    self.grab_keyboard_destroy.link.remove();
+
+    if (keyboard_grab.keyboard) |keyboard| {
+        keyboard_grab.input_method.seat.keyboardNotifyModifiers(&keyboard.modifiers);
+    }
+}
+
+pub fn getFocusableTextInput(self: *Self) ?*TextInput {
+    var it = self.text_inputs.first;
+    return while (it) |node| : (it = node.next) {
+        const text_input = &node.data;
+        if (text_input.pending_focused_surface != null) break text_input;
+    } else null;
+}
+
+pub fn getFocusedTextInput(self: *Self) ?*TextInput {
+    var it = self.text_inputs.first;
+    return while (it) |node| : (it = node.next) {
+        const text_input = &node.data;
+        if (text_input.wlr_text_input.focused_surface != null) break text_input;
+    } else null;
+}
+
+pub fn disableTextInput(self: *Self, text_input: *TextInput) void {
+    if (self.input_method) |im| {
+        im.sendDeactivate();
+    } else {
+        log.debug("disabling text input but input method is gone", .{});
+        return;
+    }
+
+    self.sendInputMethodState(text_input.wlr_text_input);
+}
+
+pub fn sendInputMethodState(self: *Self, wlr_text_input: *wlr.TextInputV3) void {
+    const input_method = self.input_method orelse return;
+
+    // TODO: only send each of those if they were modified
+    // after activation, all supported features must be sent
+
+    if (wlr_text_input.active_features.surrounding_text) {
+        if (wlr_text_input.current.surrounding.text) |text| {
+            input_method.sendSurroundingText(
+                text,
+                wlr_text_input.current.surrounding.cursor,
+                wlr_text_input.current.surrounding.anchor,
+            );
+        }
+    }
+
+    input_method.sendTextChangeCause(wlr_text_input.current.text_change_cause);
+
+    if (wlr_text_input.active_features.content_type) {
+        input_method.sendContentType(
+            wlr_text_input.current.content_type.hint,
+            wlr_text_input.current.content_type.purpose,
+        );
+    }
+
+    input_method.sendDone();
+    // TODO: pass intent, display popup size
```

Comment:
this comment is inherited from sway

Reply:
Ok I see.

Since it's still very unclear to me what "pass intent" means and I don't think cargo culting from sway is a good idea I think this comment should be removed.

If we want a hint to help the popup implementation it should be a lot more clear, something like
```
// TODO send the text_input_rectangle event here
```